tools/packet_inspector/src/lib.rs

```diff
@@ -2,99 +2,201 @@ mod packet_io;
 use valence::protocol::packets::handshaking::HandshakeC2s;
-use valence::protocol::packets::login::{LoginCompressionS2c, LoginSuccessS2c};
-use valence::protocol::{Decode, Packet as ValencePacket};
+use valence::protocol::packets::login::{
+    LoginCompressionS2c, LoginDisconnectS2c, LoginHelloS2c, LoginSuccessS2c,
+};
+use valence::protocol::{Decode, Encode, Packet as ValencePacket};
+use valence::text::color::NamedColor;
+use valence::text::{Color, IntoText};
 use valence::CompressionThreshold;
 
 use crate::packet_io::PacketIo;
 use crate::packet_registry::PacketRegistry;
-pub use crate::packet_registry::{PacketSide, PacketState};
-
-static PACKET_REGISTRY: OnceLock<Arc<RwLock<PacketRegistry>>> = OnceLock::new();
+pub use crate::packet_registry::{Packet, PacketSide, PacketState};
 
 include!(concat!(env!("OUT_DIR"), "/packets.rs"));
 
+/// Messages for talking to the running proxy task
+#[derive(Debug, Clone)]
+pub enum ProxyMessage {
+    Stop,
+    // In the future there could be a message like
+    // InjectPacket(SocketAddr, PacketFrame)
+}
+
+#[derive(Debug)]
+pub enum DisconnectionReason {
+    OnlineModeRequired,
+    Error(anyhow::Error),
+}
+
+/// Messages sent by the proxy for the controlling GUI/CLI.
+#[derive(Debug)]
+pub enum ProxyLog {
+    /// A new client has connected to the listener.
+    ClientConnected(SocketAddr),
+    /// A client has disconnected from the listener.
+    ClientDisconnected(SocketAddr, DisconnectionReason),
+}
+
 pub struct Proxy {
-    listener_addr: SocketAddr,
-    server_addr: SocketAddr,
+    pub main_task: JoinHandle<anyhow::Result<()>>,
+    pub message_tx: flume::Sender<ProxyMessage>,
+    pub logs_rx: flume::Receiver<ProxyLog>,
+    pub packet_registry: Arc<RwLock<PacketRegistry>>,
 }
 
 impl Proxy {
-    pub fn new(listener_addr: SocketAddr, server_addr: SocketAddr) -> Self {
-        PACKET_REGISTRY.get_or_init(|| {
+    /// Creates a new proxy, and starts its listener task.
+    pub async fn start(listener_addr: SocketAddr, server_addr: SocketAddr) -> anyhow::Result<Self> {
+        let (message_tx, message_rx) = flume::unbounded();
+        let (logs_tx, logs_rx) = flume::unbounded();
+
+        let packet_registry = Arc::new(RwLock::new({
             let registry = PacketRegistry::new();
             registry.register_all(&STD_PACKETS);
-            Arc::new(RwLock::new(registry))
-        });
+            registry
+        }));
 
-        Proxy {
-            listener_addr,
+        let main_task = tokio::spawn(Self::run_main_task(
+            packet_registry.clone(),
+            TcpListener::bind(listener_addr).await?,
             server_addr,
-        }
+            message_rx,
+            logs_tx,
+        ));
+
+        Ok(Self {
+            main_task,
+            message_tx,
+            logs_rx,
+            packet_registry,
+        })
     }
 
+    /// Subscribes to the proxy's [`PacketRegistry`].
```

Comment:
There's a rustdoc warning here (that should have caused CI to fail)

Reply:
It seems I can never have a day without angering rustdoc somehow huh

Either way, I just pushed a fix